File: src/WebActuator.WebAssembly/ExportManage.cs
using Microsoft.CodeAnalysis;
using Microsoft.CodeAnalysis.CSharp;
using Microsoft.JSInterop;
using System.Text.Json;
using WebActuator.Options;

namespace WebActuator.WebAssembly;

public partial class ExportManage
{
    public ExportManage()
    {
    }

    [JSInvokable]
    public static Task<string[]> Using() => Task.FromResult(GlobalManage.Using());

    [JSInvokable]
    public static async Task SetUsing(string[] usings)
    {
        GlobalManage.AddUsing(usings);
        await Task.CompletedTask;
    }

    [JSInvokable]
    public static async Task RemoveUsing(string @using)
    {
        GlobalManage.RemoveUsing(@using);
        await Task.CompletedTask;
    }

    [JSInvokable]
    public static async Task ClearUsing()
    {
        GlobalManage.ClearUsing();
        await Task.CompletedTask;
    }

    [JSInvokable]
    public static Task<int> LanguageVersion() => Task.FromResult((int)GlobalManage.LanguageVersion);

    [JSInvokable]
    public static async Task SetLanguageVersion(int languageVersion)
    {
        GlobalManage.SetLanguageVersion((LanguageVersion)languageVersion);
        await Task.CompletedTask;
    }

    [JSInvokable]
    public static Task<string[]> References() => Task.FromResult(ReferenceManage.ReferenceKeys);

    [JSInvokable]
    public static async Task SetReferencesAsync(string[] references)
    {
        foreach (var reference in references)
        {
            if (!ReferenceManage.ReferenceKeys.Contains(reference))
            {
                try
                {
                    await using var stream = await GlobalManage.HttpClient.GetStreamAsync(reference);

                    ReferenceManage.AddReference(reference, MetadataReference.CreateFromStream(stream));
                }
                catch (Exception)
                {
                }
            }
        }
    }

    [JSInvokable]
    public static Task<bool> TryCompile(string source, bool concurrentBuild = false)
    {
        return Task.FromResult(ActuatorCompile.TryCompile(new CompileOptions { ConcurrentBuild = concurrentBuild, Source = source }));
    }

    [JSInvokable]
    public static async Task RunSubmission(string code, bool concurrentBuild = false)
    {
        // await ActuatorCompile.RunSubmission(code, concurrentBuild, async (diagnostic) =>
        // {
        //     var result = JsonSerializer.Serialize(diagnostic);
        //     await WebActuatorApp.GetRequiredService<IJSRuntime>().InvokeVoidAsync("OnDiagnostic", result);
        // }, async (exception) =>
        // {
        //     await WebActuatorApp.GetRequiredService<IJSRuntime>().InvokeVoidAsync("OnException", exception.Message);
        // });
    }

}

File: examples/Examples.Web/exportManage.js

window.OnDiagnostic = (message) => {
    try {
        window.Diagnostic(message)
    } catch (error) {
    }
}

window.OnWriteLine = (value) => {
    try {
        window.WriteLine(value)
    } catch (error) {

    }
}
window.OnException = (value) => {
    try {
        window.Exception(value)
    } catch (error) {

    }
}

export async function Using() {
    return await DotNet.invokeMethodAsync('WebActuator.WebAssembly', 'Using');
}

export async function SetUsing() {
    await DotNet.invokeMethodAsync('WebActuator.WebAssembly', 'SetUsing');
}

export async function RemoveUsing(using) {
    await DotNet.invokeMethodAsync('WebActuator.WebAssembly', 'RemoveUsing', using);
}

export async function ClearUsing() {
    await DotNet.invokeMethodAsync('WebActuator.WebAssembly', 'ClearUsing');
}

export async function LanguageVersion() {
    return await DotNet.invokeMethodAsync('WebActuator.WebAssembly', 'LanguageVersion');
}

export async function SetLanguageVersion(languageVersion) {
    await DotNet.invokeMethodAsync('WebActuator.WebAssembly', 'SetLanguageVersion', languageVersion);
}

export async function References() {
    return await DotNet.invokeMethodAsync('WebActuator.WebAssembly', 'References');
}

export async function SetReferences(references) {
    await DotNet.invokeMethodAsync('WebActuator.WebAssembly', 'SetReferencesAsync', references);
}

export async function TryCompile(source, concurrentBuild) {
    await DotNet.invokeMethodAsync('WebActuator.WebAssembly', 'TryCompile', source, concurrentBuild);
}

export async function RunSubmission(code, concurrentBuild) {
    await DotNet.invokeMethodAsync('WebActuator.WebAssembly', 'RunSubmission', code, concurrentBuild);
}